staging/src/k8s.io/code-generator/cmd/validation-gen/validators/union.go

```diff
@@ -0,0 +1,165 @@
+var newUnionMembership = types.Name{Package: libValidationPkg, Name: "NewUnionMembership"}
+
+func init() {
+	AddToRegistry(InitUnionDeclarativeValidator)
+}
+
+func InitUnionDeclarativeValidator(c *generator.Context) DeclarativeValidator {
+	return &unionDeclarativeValidator{
+		universe: c.Universe,
+	}
+}
+
+type unionDeclarativeValidator struct {
+	universe types.Universe
+}
+
+const (
+	// +union and +unionDiscriminator tag are used by openapi-gen to publish x-kubernetes-union and x-kubernetes-discriminator
+	// extensions into Kubernetes published OpenAPI.
+	discriminatorTagName = "unionDiscriminator"
+	memberTagName        = "unionMember"
+)
+
+type discriminatorParams struct {
+	Union string `json:"union,omitempty"`
+}
+
+type memberParams struct {
+	Union      string `json:"union,omitempty"`
+	MemberName string `json:"memberName,omitempty"`
+}
+
+type union struct {
+	fields       []any
+	fieldMembers []any
+
+	discriminator       *string
+	discriminatorMember any
+}
+
+type unions map[string]*union
+
+func (us unions) getOrCreate(name string) *union {
+	var u *union
+	var ok bool
+	if u, ok = us[name]; !ok {
+		u = &union{}
+		us[name] = u
+	}
+	return u
+}
+
+func (c *unionDeclarativeValidator) ExtractValidations(t *types.Type, comments []string) (ValidatorGen, error) {
+	result := ValidatorGen{}
+
+	// TODO: handle multiple unions
+	// This really only works if we name each union
+	// so the most obvious approach is to track data by union
+	unions := unions{}
+
+	for _, member := range t.Members {
+		commentTags := gengo.ExtractCommentTags("+", member.CommentLines)
+		if tags, ok := commentTags[memberTagName]; ok {
+			if len(tags) != 1 {
+				return result, fmt.Errorf("must have one %q tag", memberTagName)
+			}
+			tag := tags[0]
+			var fieldName string
+			if jsonAnnotation, ok := annotations.LookupJSONTags(member); ok {
+				fieldName = jsonAnnotation.Name
+				if len(fieldName) > 0 {
+					p := &memberParams{MemberName: member.Name}
+					if len(tag) > 0 {
+						if err := json.Unmarshal([]byte(tag), &p); err != nil {
+							return result, fmt.Errorf("error parsing JSON value: %v (%q)", err, tag)
+						}
+					}
+					u := unions.getOrCreate(p.Union)
+
+					if fieldName == strings.ToLower(p.MemberName[:1])+p.MemberName[1:] {
+						u.fields = append(u.fields, fieldName) // member name follows conventions, only track field name
+					} else {
+						u.fields = append(u.fields, [2]string{fieldName, p.MemberName}) // member name is custom, track it
+					}
+					u.fieldMembers = append(u.fieldMembers, member)
+				}
+			}
+		}
+		if tags, ok := commentTags[discriminatorTagName]; ok {
+			if len(tags) != 1 {
+				return result, fmt.Errorf("must have one %q tag", memberTagName)
+			}
+			tag := tags[0]
+
+			p := &discriminatorParams{}
+			if len(tag) > 0 {
+				if err := json.Unmarshal([]byte(tag), &p); err != nil {
```

Comment:
I like this idea in general.

My only hesitation with union name is that it's for a feature we expect to rarely use-- having multiple unions in the same go struct.  We don't have any APIs in core that need this..

Reply:
Can we make what `validateTrue` does be "normal"?  Start with no value: `+validateTrue`.  Eventually we want a string or some other value: `+validateTrue="optional value"`.  Eventually we want something more complex: `+validateTrue={"msg": "optional value", "flags": [ "flag", "flag" ]}`.

We can document that and I think it will be OK